Nearest neighbour interactions for `HubbardRealSpace`

Skuwar1 opened the pull request:
# New Features

- Added nearest neighbour interactions to `HubbardRealSpace`
- Generalized hopping parameter `t` in `HubbardRealSpace` to have directional axis-wise magnitude, i.e., for single component: `t = [t_x t_y t_z...]`
- `HubbardRealSpace` now accepts complex hopping amplitudes

Skuwar1 committed:
added nearest neighbour interaction to HubbardRealSpace

Skuwar1 committed:
Update HubbardRealSpace.jl

Skuwar1 committed:
Update HubbardRealSpace.jl

mtsch commented on src/Hamiltonians/HubbardRealSpace.jl:
```diff
@@ -310,40 +396,40 @@ dimension(::HubbardRealSpace, address) = number_conserving_dimension(address)
 # offdiaonals =========================================================================== #
 # Holds the offdiagonals for a single-component nearest neighbour one-body term. It's
-# structured like a matrix where the first index determines the occupied site in the address
+# structured like a matric where the first index determines the occupied site in the adress
```
```suggestion
# structured like a matrix where the first index determines the occupied site in the address
```

mtsch commented on src/Hamiltonians/HubbardRealSpace.jl:
```diff
@@ -53,58 +53,119 @@ end
 for (i, a) in enumerate(addresses)
-    acc += local_interaction(a, interaction_matrix[i, i])
+    acc += local_interaction(a, onsite_int_matrix[i, i]) +
+        nearest_neighbour_int_matrix(a, nearest_neighbour_int_matrix[i,i],
+        occupied_mode_map(a), geometry)
     for (j, b) in enumerate(addresses[i+1:end])
-        acc += local_interaction(a, b, interaction_matrix[i, j])
+        acc += local_interaction(a, b, onsite_int_matrix[i, j]) +
+            nearest_neighbour_int_matrix(a, b, nearest_neighbour_int_matrix[i,j],
+            occupied_mode_map(a), geometry)
```
Github won't let me suggest changes here, but you should replace `nearest_neighbor_int_matrix` with `nearest_neighbor_interaction`. Also indent it like you would in code.

mtsch commented on src/Hamiltonians/HubbardRealSpace.jl:
```diff
@@ -53,58 +53,119 @@ end
+    end
+    return Δ*ext_result
+end
+@inline nearest_neighbour_interaction(::SVector, ::SVector, ::Nothing, ::CubicGrid, ::ModeMap) = 0
```
```suggestion
@inline nearest_neighbor_interaction(::SVector, ::SVector, ::Nothing, ::CubicGrid, ::ModeMap) = 0
```
Unfortunately, we use American spelling in the code 😄

mtsch reviewed (changes requested):
Cool new features!
Some surface-level comments for now, I'll dig deeper into the code later and let you know.
Please add some tests that check if complex hopping, different values of `t` per dimension, nearest neighbour interactions work correctly.

The failing tests seem to be unrelated to the PR.

Skuwar1 committed:
Update src/Hamiltonians/HubbardRealSpace.jl

Co-authored-by: mtsch <matijacufar@gmail.com>

Skuwar1 committed:
Update HubbardRealSpace.jl

Skuwar1 committed:
Update Hamiltonians.jl

- Added test for ``nearest_neighbour_interaction`` and complex hopping in ``HubbardRealSpace``

joachimbrand commented on src/Hamiltonians/HubbardRealSpace.jl:
```diff
@@ -141,15 +202,16 @@ end
 # struct ================================================================================ #
 """
-    HubbardRealSpace(address; geometry=PeriodicBoundaries(M,), t=ones(C), u=ones(C, C), v=zeros(C, D))
+    HubbardRealSpace(address; geometry=PeriodicBoundaries(M,), t=ones(C, D), u=ones(C, C), Δ=zeros(C, C), v=zeros(C, D))
```
```suggestion
    HubbardRealSpace(address; geometry=PeriodicBoundaries(M,), t=ones(C, D), u=ones(C, C), v=zeros(C, D), vprime=zeros(C, C))
```

I'm not in favour of naming the interaction parameter \Delta for two reasons: For one it is better to avoid unicode names for arguments in library functions. The other reason is that \Delta is already used for a reduced interaction in the (mapped) spin chain models (and if I understand correctly the value is different). My suggestion `vprime` at least has a precedent that it was used in the literature before (for the t-V-V' model).

joachimbrand commented on src/Hamiltonians/HubbardRealSpace.jl:
```diff
@@ -168,7 +230,7 @@ along dimension ``d``.
 
 Note that a single component of fermions cannot interact with itself. A warning
-is produced if `address`is incompatible with the interaction parameters `u`.
+is produced if `address`is incompatible with the interaction parameters `u` and `Δ`.
```
Why would we need a warning and what incompatibility could there be between an address and nearest neighbour interaction? The Pauli principle doesn't forbid particles interacting with the same component by this interaction.

joachimbrand commented on src/Hamiltonians/HubbardRealSpace.jl:
```diff
@@ -141,15 +202,16 @@ end
-  \\frac{1}{2}\\sum_{i,σ} u_{σσ} n_{iσ} (n_{iσ} - 1) +
+  \\hat{H} = -\\sum_{\\langle i,j\\rangle,σ} t_{iσ} a^†_{iσ} a_{jσ} +
+  \\frac{1}{2}\\sum_{i,σ,σ'} u_{σσ'} n_{iσ} (n_{iσ'} - δ_{σ,σ'}) + 
+  \\sum_{⟨i,j⟩,σ,σ'} Δ_{iσ,jσ'} n_{iσ} n_{jσ} +
```
From the formula it looks like  Δ can vary across the lattice (it depends on the site indices and the component indices), but the function only accepts a D by D matrix. So should this be the following instead?
```suggestion
  \\sum_{⟨i,j⟩,σ,σ'} V'_{σ,σ'} n_{iσ} n_{jσ} +
```

joachimbrand reviewed (changes requested):
This looks like a really nice feature extension. 
Only a couple of quick comments so far - I'll need to review this more thoroughly later.

mtsch committed:
Clean up constructor, add tests

mtsch committed:
fix typo in docstring

mtsch committed:
change type of P

Skuwar1 committed:
Update HubbardRealSpace.jl

Skuwar1 committed:
Update Hamiltonians.jl

mtsch commented on src/Hamiltonians/HubbardRealSpace.jl:
```diff
@@ -183,85 +245,121 @@ number of sites `M` inferred from the number of modes in `address`.
     address::A
     t::T # hopping strengths
     u::U # interactions
+    w::X # nearest neighbour interactions
```
```suggestion
    w::W # nearest neighbour interactions
```

Skuwar1 committed:
Update src/Hamiltonians/HubbardRealSpace.jl

Co-authored-by: mtsch <matijacufar@gmail.com>

mtsch commented on src/Hamiltonians/HubbardRealSpace.jl:
```diff
@@ -53,58 +53,119 @@ end
+"""
+@inline function nearest_neighbor_interaction(onr1::SVector, onr2::SVector,
+    w, geometry::CubicGrid{D,S,B}, map1::ModeMap
+    ) where {D,S,B}
```
```suggestion
@inline function nearest_neighbour_interaction(
    onr1::SVector, onr2::SVector, Δ, geometry::CubicGrid{D,S,B}, map1::ModeMap
) where {D,S,B}
```
It's better to break long lines like this. See [this style guide](https://github.com/JuliaDiff/BlueStyle?tab=readme-ov-file#method-definitions). Also please go through and make sure that your lines are not over 92 characters long. It makes the code easier to read.

mtsch commented on src/Hamiltonians/HubbardRealSpace.jl:
```diff
@@ -53,58 +53,119 @@ end
+end
+@inline nearest_neighbor_interaction(::SVector, ::SVector, ::Nothing, ::CubicGrid, ::ModeMap) = 0
+
+@inline _interactions_(addr::SingleComponentFockAddress, ::Nothing, w, occs,
```
This is confusing. Why are there two functions, `_interactions_` and `_interactions`?
We also generally don't have functions with an underscore at the end.

mtsch commented on src/Hamiltonians/HubbardRealSpace.jl:
```diff
@@ -53,58 +53,119 @@ end
+@inline _interactions(::Tuple{}, ::Union{SMatrix{0,0},Nothing}, ::Union{SMatrix{0,0},Nothing},
+    ::Tuple{}, ::CubicGrid) = 0.0
+@inline function _interactions((a, as...)::NTuple{N,AbstractFockAddress},
+    m::Union{SMatrix{N,N},Nothing}, σ::Union{SMatrix{N,N},Nothing}, (occ, occs...),
```
```suggestion
    m::Union{SMatrix{N,N},Nothing}, σ::Union{SMatrix{N,N},Nothing}, (occ, occs...),
```
I don't know why I originally named the matrix `m` (looks like a bad idea to me). It'd be better if these were `u_matrix` and `w_matrix`, or something similar so it's clearer what they represent.

mtsch commented on src/Hamiltonians/HubbardRealSpace.jl:
```diff
@@ -53,58 +53,119 @@ end
-    (u, column...) = Tuple(m[:, 1])
+    (u, u_column...) = isnothing(m) ? (nothing, nothing) : Tuple(m[:, 1])
+    (w, w_column...) = isnothing(σ) ? (nothing, nothing) : Tuple(σ[:, 1])
     # Type-stable way to subset SMatrix:
-    rest = SMatrix{N-1,N-1}(view(m, 2:N, 2:N))
-
+    m_rest = isnothing(m) ? nothing : SMatrix{N-1,N-1}(view(m, 2:N, 2:N))
+    σ_rest = isnothing(σ) ? nothing : SMatrix{N-1,N-1}(view(σ, 2:N, 2:N))
```
This is good, but it might be even better to split the logic into a function with a descriptive name. The function would take `m` and return `u`, `u_column` and `m_rest`. You can then use dispatch to handle the `nothing`s and call it twice for `m` and `σ`.

mtsch reviewed (changes requested):
Some more comments, mostly regarding readability. 
Please apply the multiline function definition comment everywhere.

Skuwar1 committed:
Update HubbardRealSpace.jl

Skuwar1 committed:
Update HubbardRealSpace.jl

Skuwar1 committed:
Update HubbardRealSpace.jl

Skuwar1 committed:
Update HubbardRealSpace.jl

Skuwar1 committed:
Update HubbardRealSpace.jl

Skuwar1 committed:
Update HubbardRealSpace.jl

Skuwar1 committed:
Update HubbardRealSpace.jl

Skuwar1 committed:
Update HubbardRealSpace.jl

Skuwar1 committed:
Update HubbardRealSpace.jl

Skuwar1 committed:
Update HubbardRealSpace.jl

mtsch committed:
formatting

mtsch reviewed (approved):
I made some formatting edits. I think it's ready to be merged.

mtsch commented:
I think this PR solves #293

mtsch commented on src/Hamiltonians/HubbardRealSpace.jl:
```diff
@@ -183,126 +319,191 @@ number of sites `M` inferred from the number of modes in `address`.
     C = num_components(address)
+    U = Matrix(u)
     for c in 1:C
-        if address.components[c] isa FermiFS && u ≠ ones(C,C) && u[c,c] ≠ 0
+        if address.components[c] isa FermiFS && u[c, c] ≠ 0
             @warn "component $(c) is fermionic, but was given a self-interaction " *
                 "strength of $(u[c,c])" maxlog=1
+            U[c, c] = 0
         end
     end
+    return SMatrix{size(U)...,eltype(u)}(U)
 end
 function warn_fermi_interaction(address::FermiFS, u)
-    if u ≠ ones(1, 1) && u[1, 1] ≠ 0
+    if u[1, 1] ≠ 0
         @warn "address is fermionic, but was given a self-interaction " *
             "strength of $(u[1,1])" maxlog=1
     end
+    return SMatrix{1,1,eltype(u)}(0)
 end
```
I just noticed the tests are failing. They started failing when this was changed. I'd suggest reverting it back, setting `U[c, c]` to zero doesn't do anything here (since the fermions don't interact either way), and the `u ≠ ones(C,C)` is there to not trigger the warning when using the default interaction of 1.

joachimbrand commented on src/Hamiltonians/HubbardRealSpace.jl:
```diff
@@ -141,24 +275,26 @@ end
+  Ĥ = -∑_{⟨i,j⟩,σ} t_{i,σ} a^†_{i,σ} a_{j,σ} +
+  \\frac{1}{2} ∑_{i,σ,σ'} u_{σ,σ'} n_{i,σ} (n_{i,σ'} - w_{σ,σ'}) +
+  ∑_{⟨i,j⟩,σ,σ'} w_{i,σ,j,σ'} n_{i,σ} n_{j,σ} +
+  ∑_{i,σ≠τ} u_{σ,τ} n_{i,σ} n_{i,τ}
```
```suggestion
  Ĥ = -∑_{⟨i,j⟩,σ} t_{i,σ} a^†_{i,σ} a_{j,σ} +
  \\frac{1}{2} ∑_{i,σ,σ'} u_{σ,σ'} n_{i,σ} (n_{i,σ'} - 1) +
  ∑_{⟨i,j⟩,σ,σ'} w_{σ,σ'} n_{i,σ} n_{j,σ}
```
The formula was incorrect. My suggestion is what I think it should be. But please double check that it makes sense.

joachimbrand commented on src/Hamiltonians/HubbardRealSpace.jl:
```diff
@@ -183,126 +319,191 @@ number of sites `M` inferred from the number of modes in `address`.
 * `u`: the on-site interaction parameters. Must be a symmetric matrix. `u[i, j]`
   corresponds to the interaction between the `i`-th and `j`-th component. `u[i, i]`
   corresponds to the interaction of a component with itself. Note that `u[i,i]` must
   be zero for fermionic components.
```
```suggestion
* `u`: the on-site interaction parameters. Must be a symmetric matrix of size
  `C × C`. `u[i, j]` corresponds to the interaction between the `i`-th and `j`-th 
  component. `u[i, i]` corresponds to the interaction of a component with itself.
  Note that the self interaction will have no effect on fermionic components.
```

joachimbrand commented on src/Hamiltonians/HubbardRealSpace.jl:
```diff
@@ -183,126 +319,191 @@ number of sites `M` inferred from the number of modes in `address`.
   be zero for fermionic components.
+* `w`: the nearest neighbour interaction parameters. Must be a symmetric matrix. `w[i, j]`
+  corresponds to the interaction between the `i`-th and `j`-th component. `w[i, i]`
+  corresponds to the interaction of a component with itself.
```
```suggestion
* `w`: the nearest neighbour interaction parameters. Must be a symmetric matrix of size
  `C × C`. `w[i, j]` corresponds to the interaction between the `i`-th and `j`-th component.
   `w[i, i]` corresponds to the interaction of a component with itself.
```

joachimbrand reviewed (changes requested):
An incomplete review for now, as the test are failing.

I noticed a problem. It might be related to Matija's comment or the tests failing:
```julia-repl
julia> h = HubbardRealSpace(BoseFS(1,2,0,0,2,3); u=2.0)
HubbardRealSpace(
  fs"|1 2 0 0 2 3⟩",
  geometry = CubicGrid((6,), (true,)),
  t = [1.0;;],
  u = 0.0
)
```
For some reason a nonzero on-site interaction is not accepted for this bosonic address.

I'm also wondering whether `warn_fermi_interactions` is needed at all. If people set the self-interaction of a fermionic component to a non-zero value, why should we care? It has no effect, but there is not logical problem with setting it. 

Please re-request a code review after the tests are passing (or failing tests are clearly unrelated to your changes)!

Skuwar1 committed:
Update src/Hamiltonians/HubbardRealSpace.jl

Co-authored-by: Joachim Brand <joachim.brand@gmail.com>

Skuwar1 committed:
Update src/Hamiltonians/HubbardRealSpace.jl

Co-authored-by: Joachim Brand <joachim.brand@gmail.com>

Skuwar1 committed:
Update src/Hamiltonians/HubbardRealSpace.jl

Co-authored-by: Joachim Brand <joachim.brand@gmail.com>

Skuwar1 committed:
Update HubbardRealSpace.jl

mtsch commented on src/Hamiltonians/HubbardRealSpace.jl:
```diff
@@ -183,126 +318,187 @@ number of sites `M` inferred from the number of modes in `address`.
     C = num_components(address)
     for c in 1:C
-        if address.components[c] isa FermiFS && u ≠ ones(C,C) && u[c,c] ≠ 0
+        if address.components[c] isa FermiFS && u[c, c] ≠ 0
```
There used to be a check here that didn't print the warning if `u` was at its default value. I don't think this should be printed when the user doesn't pass a value of `u` to the constructor.

mtsch commented on src/Hamiltonians/HubbardRealSpace.jl:
```diff
@@ -183,126 +318,187 @@ number of sites `M` inferred from the number of modes in `address`.
     warn_fermi_interaction(address, u)
 
-Warn if interaction matrix `u` does not make sense for `address`.
+Warn and modify if interaction matrix `u` does not make sense for `address`.
```
```suggestion
Warn if interaction matrix `u` does not make sense for `address`.
```

mtsch commented on src/Hamiltonians/HubbardRealSpace.jl:
```diff
@@ -183,126 +318,187 @@ number of sites `M` inferred from the number of modes in `address`.
 end
 function warn_fermi_interaction(address::FermiFS, u)
-    if u ≠ ones(1, 1) && u[1, 1] ≠ 0
+    if u[1, 1] ≠ 0
```
Same here.

I think it's also OK to just remove this check altogether.

mtsch commented on src/Hamiltonians/HubbardRealSpace.jl:
```diff
@@ -183,126 +318,187 @@ number of sites `M` inferred from the number of modes in `address`.
 
-    t_vec = SVector{C,Float64}(t)
-    u_mat = iszero(u) ? nothing : SMatrix{C,C,Float64}(u)
+    warn_fermi_interaction(address, u_mat)
```
As an alternative you could do the check like this and not modify `warn_fermi_interction`.
```suggestion
    u ≢ 1.0 && warn_fermi_interaction(address, u_mat)
```

Skuwar1 committed:
Update src/Hamiltonians/HubbardRealSpace.jl

Co-authored-by: mtsch <matijacufar@gmail.com>

Skuwar1 committed:
Update src/Hamiltonians/HubbardRealSpace.jl

Co-authored-by: mtsch <matijacufar@gmail.com>

Skuwar1 committed:
Update HubbardRealSpace.jl

joachimbrand commented on src/Hamiltonians/HubbardRealSpace.jl:
```diff
@@ -141,24 +275,25 @@ end
-  \\sum_{i,σ≠τ}u_{στ} n_{iσ} n_{iτ}
+  Ĥ = -∑_{⟨i,j⟩,σ} t_{i,σ} a^†_{i,σ} a_{j,σ} +
+  \\frac{1}{2} ∑_{i,σ,σ'} u_{σ,σ'} n_{i,σ} (n_{i,σ'} - 1) +
+  ∑_{⟨i,j⟩,σ,σ'} w_{σ,σ'} n_{i,σ} n_{j,σ}
```
```suggestion
  ∑_{⟨i,j⟩,σ,σ'} w_{σ,σ'} n_{i,σ} n_{j,σ'}
```

joachimbrand commented on src/Hamiltonians/HubbardRealSpace.jl:
```diff
@@ -53,58 +53,192 @@ end
+
+Sum the local interactions of the Fock state `a` with all states in `bs` using the onsite
+and nearest neighbour interaction constants in `us` and `ws`. This is used to compute
+all interactions in the column below the diagonal of the interaction matrix.
```
This is a little opaque on first reading. Can you please make it clearer what the roles of the arguments are and what types are to be passed? Maybe having longer and more meaningful names for the variables would already help a lot. E.g. is `occ` expected to be the `occupied_mode_map` of `a` and `occs` for `bs`?

Skuwar1 replied:
Have done changes.

joachimbrand commented on src/Hamiltonians/HubbardRealSpace.jl:
```diff
@@ -53,58 +53,192 @@ end
+)
+    return local_interaction(a, b, u, occ_a, occ_b) +
+           _interaction_col(a, bs, us, ws, occ_a, occs, g) +
+           nearest_neighbor_interaction(onr(a), onr(b), w, g, occ_a)
```
I'm a little worried about calling `onr` here (and the time it take, even if it is little), where the relevant information in the `occupied_mode_map`s is already there. It might be better to pass the `ModeMap`s into `nearest_neighbor_interaction` and define a method that accepts those?

mtsch replied:
Here, you need to check whether a site is occupied, so I think using `onr` is the best option here.

Skuwar1 replied:
Have done some changes. `onr` in the argument is now replaced with `address`

joachimbrand replied:
Ok, I agree that it was done the right way. It doesn't matter whether `onr` is called here or inside the function. I thought that there might be a way to speed it up, but there isn't, so don't worry about it.

joachimbrand commented on src/Hamiltonians/HubbardRealSpace.jl:
```diff
@@ -53,58 +53,192 @@ end
+    w, w_column, w_rest = _dismantle_int_matrix(w_matrix)
     # Get the self-interaction first.
-    self = local_interaction(a, u, occ)
+    self = local_interaction(a, u, occ) + nearest_neighbor_interaction(onr(a), w, g, occ)
```
Same as above. We shouldn't need to call `onr` here.

joachimbrand commented on src/Hamiltonians/HubbardRealSpace.jl:
```diff
@@ -53,58 +53,192 @@ end
 """
-    _interaction_col(a, bs::Tuple, us::Tuple, occ::ModeMap, occs::Tuple)
+    nearest_neighbor_interaction(onr::SVector, w, geometry::CubicGrid, map::ModeMap)
+    nearest_neighbor_interaction(onr1::SVector, onr2::SVector, w, geometry::CubicGrid, map1::ModeMap)
```
Please clarify in the docstring what `ModeMap` should be passed. 

Maybe we can generate the `onr` from `ModeMap`s faster than from the bitstring address?

mtsch replied:
I was worried about that as well and I benchmarked it. Depending on the address, it's sometimes actually faster to do it directly from the address.

joachimbrand replied:
OK, I tried and failed to generate the `onr` from the `ModeMap` faster than from the address. Let's just leave it as is.

joachimbrand reviewed (changes requested):
A few small things. See below.

Skuwar1 committed:
Update src/Hamiltonians/HubbardRealSpace.jl

Co-authored-by: Joachim Brand <joachim.brand@gmail.com>

Skuwar1 committed:
Update HubbardRealSpace.jl

Skuwar1 committed:
Update HubbardRealSpace.jl

joachimbrand commented on src/Hamiltonians/HubbardRealSpace.jl:
```diff
@@ -53,58 +53,200 @@ end
+
+Calculate the nearest neighbour interaction ``w \\sum_{⟨i,j⟩} n_{↑, i} n_{↓, j}`` within a same single-component
+Fock address `add` or between the single-component Fock states `add1` and `add2` of two distinct components in 
+multi-component Fock address,
```
```suggestion
Calculate the nearest neighbour interaction ``w \\sum_{⟨i,j⟩} n_{↑, i} n_{↓, j}`` within 
the single-component Fock address `add`. When two single-component Fock states 
`add1` and `add2` are passed, return the eigenvalue of
```

joachimbrand commented on src/Hamiltonians/HubbardRealSpace.jl:
```diff
@@ -53,58 +53,200 @@ end
-the intraction strengths and `occs` holds the occupied modes of the adresses.
+Compute all pairwise interactions in a tuple of `addresses`. The `onsite_int_matrix` and
+`nearest_neighbour_int_matrix` sets the intraction strengths of the onsite interaction and
+the nearest neighbour interaction. Moreover, `occs` holds the occupied modes of the adresses
```
```suggestion
the nearest neighbour interaction. Moreover, `occs` holds the occupied modes of the adresses.
```

joachimbrand commented on src/Hamiltonians/HubbardRealSpace.jl:
```diff
@@ -141,24 +283,25 @@ end
-  \\hat{H} = -\\sum_{\\langle i,j\\rangle,σ} t_σ a^†_{iσ} a_{jσ} +
-  \\frac{1}{2}\\sum_{i,σ} u_{σσ} n_{iσ} (n_{iσ} - 1) +
-  \\sum_{i,σ≠τ}u_{στ} n_{iσ} n_{iτ}
+  Ĥ = -∑_{⟨i,j⟩,σ} t_{i,σ} a^†_{i,σ} a_{j,σ} +
+  \\frac{1}{2} ∑_{i,σ,σ'} u_{σ,σ'} n_{i,σ} (n_{i,σ'} - 1) +
+  ∑_{⟨i,j⟩,σ,σ'} w_{σ,σ'} n_{i,σ} n_{j,σ'}
```
The contributions to the on-site interactions from different components are not handled correctly. 
Please double check that the code implementation is correct!

```suggestion
  Ĥ = -∑_{⟨i,j⟩,σ} t_{i,σ} a^†_{i,σ} a_{j,σ} +
   ½ ∑_{i,σ,σ'} u_{σ,σ'} n_{i,σ} (n_{i,σ'} - δ_{σ,σ'}) +
  ∑_{⟨i,j⟩,σ,σ'} w_{σ,σ'} n_{i,σ} n_{j,σ'}
```

joachimbrand commented on src/Hamiltonians/HubbardRealSpace.jl:
```diff
@@ -283,26 +482,31 @@ function warn_fermi_interaction(address::FermiFS, u)
+function Base.show(io::IO, h::HubbardRealSpace{TT,C}) where {TT,C}
     io = IOContext(io, :compact => true)
     println(io, "HubbardRealSpace(")
     println(io, "  ", starting_address(h), ",")
```
```suggestion
    println(io, "  ", starting_address(h), ";")
```

joachimbrand commented on test/Hamiltonians.jl:
```diff
@@ -38,13 +38,13 @@ end
         ExtendedHubbardMom1D(OccupationNumberFS(1,2,0,0); u=1.0, v=2.0, t=3.0),
         ExtendedHubbardMom1D(FermiFS(1,1,0,0); u=1.0, v=2.0, t=3.0),
-        HubbardRealSpace(BoseFS((1, 2, 3)); u=[1], t=[3]),
+        HubbardRealSpace(BoseFS((1, 2, 3)); u=[1], t=[3], w=[1]),
```
Can we add `HubbardRealSpace` with complex hopping parameters here as well in the interface tests?

joachimbrand reviewed (changes requested):
Looks mostly good to me. Please see comments below!

Skuwar1 committed:
Update src/Hamiltonians/HubbardRealSpace.jl

Co-authored-by: Joachim Brand <joachim.brand@gmail.com>

Skuwar1 committed:
Update src/Hamiltonians/HubbardRealSpace.jl

Co-authored-by: Joachim Brand <joachim.brand@gmail.com>

Skuwar1 committed:
Update src/Hamiltonians/HubbardRealSpace.jl

Co-authored-by: Joachim Brand <joachim.brand@gmail.com>

Skuwar1 committed:
Update Hamiltonians.jl

-Add interface test for ``HubbardRealSpace`` with complex hopping.

Skuwar1 committed:
Update src/Hamiltonians/HubbardRealSpace.jl

Co-authored-by: Joachim Brand <joachim.brand@gmail.com>

Skuwar1 committed:
Update Hamiltonians.jl

joachimbrand reviewed (approved):
Thanks for the PR! All looks good now.
